### PR TITLE
[LTB-49] Remove manual connection handling

### PR DESCRIPTION
**Pull request by volhovm**
See commit messages.

### CODE REVIEW

File: code/network/lib/Loot/Network/ZMQ/Client.hs [hunk 1]
```diff
@@ -307,105 +262,95 @@ createNetCliEnv globalEnv@ZTGlobalEnv{..} ztCliSettings peers = liftIO $ do
...
+        let both = L.nub $ _uprDel `L.intersect` _uprAdd
+        let add' = L.nub _uprAdd L.\\ both
+        let del' = L.nub _uprDel L.\\ both
+
+        currentPeers <- HSet.toList . hmKeys <$> readTVar piv
+
+        -- We're adding all peers asked which were not connected.
+        -- And deleting all who are not connected.
+        let toAdd = add' L.\\ currentPeers
+        let toDel = del' L.\\ currentPeers
```
Comment (pva701) [inline code/network/lib/Loot/Network/ZMQ/Client.hs]: lol, it seems to be an error.
if `currentPeers == S.fromList [ZTNodeId "misha.allo.ru" 100 200]` and you ask to disconnect it, after this line `toDel` will be empty, though, you had to disconnect this peer.

it probably should be `L.intersect`

File: code/network/lib/Loot/Network/ZMQ/Client.hs [hunk 2]
```diff
@@ -433,75 +381,37 @@ reconnectPeers ZTNetCliEnv{..} nIds = liftIO $ do
...
-                "Connecting attempt timeouted for peers: " <>
-                show (map fst disconnected)
-            forM_ disconnected $ \(_,(_,sock,adapter)) ->
-                Z.close sock >> adapterRelease adapter
-    handler e = do
-        ztLog ztCliLogging Error $
-            "Connections timeout worker failed: " <> show e <>
-            ", restarting in 2s"
-        threadDelay 2000000
+    forM_ toReconnect $ \(nId, PeerRes{..}) -> do
```
Comment (pva701) [inline code/network/lib/Loot/Network/ZMQ/Client.hs]: shouldn't adapters be somehow affected after reconnecting?

File: code/network/lib/Loot/Network/ZMQ/Client.hs [hunk 3]
```diff
@@ -433,75 +381,37 @@ reconnectPeers ZTNetCliEnv{..} nIds = liftIO $ do
...
+        Z.disconnect prBack (ztNodeIdRouter nId)
```
Comment (pva701) [inline code/network/lib/Loot/Network/ZMQ/Client.hs]: honestly I would prefer to disconnect both and then connect both and do it atomically to avoid inconsistency between two sockets. thought, it's not so important on current stage of development, IMHO

File: code/network/lib/Loot/Network/ZMQ/Client.hs [hunk 4]
```diff
@@ -433,75 +381,37 @@ reconnectPeers ZTNetCliEnv{..} nIds = liftIO $ do
...
     unless (null toReconnect) $
         ztLog ztCliLogging Warning $ "Reconnecting peers: " <> show nIds
-    forM_ toReconnect $ \nId -> do
-        Z.disconnect ztCliBack (ztNodeIdRouter nId)
-        Z.connect ztCliBack (ztNodeIdRouter nId)
-        Z.disconnect ztCliSub (ztNodeIdPub nId)
-        Z.connect ztCliSub (ztNodeIdPub nId)
-
--- This worker cleans up connection attempts that were stuck.
-connectionsWorker :: ZTNetCliEnv -> IO ()
```
Comment (pva701) [inline code/network/lib/Loot/Network/ZMQ/Client.hs]: finally, it's remained to remove heartbeat worker :peka2:

File: code/network/lib/Loot/Network/ZMQ/Client.hs [hunk 5]
```diff
@@ -558,92 +468,72 @@ runBroker = do
...
+            whenJust nodeId' $ \nodeId ->
+                atomically (HMap.lookup nodeId <$> readTVar ztPeers) >>= \case
+                    Nothing ->
+                        ztCliLog Warning $
+                        "Coludn't send to peer, can't resolve: " <> show nodeId
+                    Just PeerRes{ prBack } -> do
+                        Z.sendMulti prBack $ NE.fromList $
+                            ["", unMsgType msgT] ++ msg
+
+    let routerToClients nodeId (MsgType -> msgT) msg = do
```
Comment (pva701) [inline code/network/lib/Loot/Network/ZMQ/Client.hs]: lol, there is no `ROUTER` anymore :peka2:

File: code/network/lib/Loot/Network/ZMQ/Client.hs [hunk 6]
```diff
@@ -558,92 +468,72 @@ runBroker = do
...
-              Nothing -> ztCliLog Warning $ "subToClients: can't resolve iId" <> show iId
+            nodeId' <- maybe choosePeer (pure . Just) nodeIdM
+            whenJust nodeId' $ \nodeId ->
+                atomically (HMap.lookup nodeId <$> readTVar ztPeers) >>= \case
+                    Nothing ->
+                        ztCliLog Warning $
+                        "Coludn't send to peer, can't resolve: " <> show nodeId
+                    Just PeerRes{ prBack } -> do
+                        Z.sendMulti prBack $ NE.fromList $
+                            ["", unMsgType msgT] ++ msg
```
Comment (pva701) [inline code/network/lib/Loot/Network/ZMQ/Client.hs]: is this `""` really needed to send?

File: code/network/lib/Loot/Network/ZMQ/Client.hs [hunk 7]
```diff
@@ -654,30 +544,30 @@ runBroker = do
...
-                          map (\(nId, (_,sock,adapter)) ->
-                                boolToMaybe (CBDealer nId sock) <$>
-                                adapterTry adapter)
-                              connecting
-                  atLeastOne $ NE.fromList $
-                      [ readReq
-                      , boolToMaybe CBSub <$> adapterTry ztCliSubAdapter
-                      , boolToMaybe CBBack <$> adapterTry ztCliBackAdapter
-                      ] ++ readClient ++ readDealers
+                  let readBacksSubs =
```
Comment (pva701) [inline code/network/lib/Loot/Network/ZMQ/Client.hs]: it's separate function now :pogchamp:

File: code/network/lib/Loot/Network/ZMQ/Client.hs (hunk 6 above)
Comment (pva701) [reply code/network/lib/Loot/Network/ZMQ/Client.hs]: it seems that server also expects `clientId` as first frame, doesn't it?

File: code/network/lib/Loot/Network/ZMQ/Client.hs [hunk 8]
```diff
@@ -191,37 +183,84 @@ data InternalRequest
...
+    pure CliPollData{..}
+
+-- | This queue is used to transfer messages from clients to the main broker.
+type ClientsQueue = InternalQueue (Maybe ZTNodeId, (MsgType, Content))
+
+-- | This worker reads from clients' bitqueues and forwards these
+-- messages to one socket that is then processed by broker in one big
+-- zmq poll.
+clientsToBrokerWorker :: ZTNetCliEnv -> IO ()
+clientsToBrokerWorker ZTNetCliEnv { ztClientsQueue, ztClients, ztCliLogging } =
```
Comment (pva701) [inline code/network/lib/Loot/Network/ZMQ/Client.hs]: nice copy-paste from `listenersToWorker`

does it make sense to abstract it in `Common.hs`

File: code/network/lib/Loot/Network/ZMQ/Client.hs [hunk 9]
```diff
@@ -527,47 +577,31 @@ runBroker = do
...
+              CliPollData{..} <- atomically $ readTVar ztCliPollData
+
+              let resolve i =
+                      fromMaybe (error $ "lookupUnsafe " <> show (i,cpdN)) $
+                      HMap.lookup i cpdResolveMap
+
+              events <- Z.poll (-1) cpdPolls
+              forM_ (events `zip` [0..]) $ \(e,i) ->
+                  unless (null e) $
+                      if | i < cpdN ->
```
Comment (pva701) [inline code/network/lib/Loot/Network/ZMQ/Client.hs]: honestly, I'd prefer to have separate function `classifySocketIdx` which returns the tag by socket index

File: code/network/lib/Loot/Network/ZMQ/Client.hs [hunk 10]
```diff
@@ -191,37 +183,84 @@ data InternalRequest
...
+----------------------------------------------------------------------------
+-- Polling data
+----------------------------------------------------------------------------
 
--- | Wrapper over client request queue. CLient broker listens to this.
-newtype CliRequestQueue = CliRequestQueue
-    { unCliRequestQueue :: TQueue InternalRequest
+data CliPollData = CliPollData
+    { cpdPolls      :: ![Z.Poll Z.Socket IO]
+    , cpdN          :: !Int
```
Comment (pva701) [inline code/network/lib/Loot/Network/ZMQ/Client.hs]: can't `cpdN` be computed as `length cpdPolls`?

File: code/network/lib/Loot/Network/ZMQ/Client.hs (hunk 2 above)
Comment (pva701) [reply code/network/lib/Loot/Network/ZMQ/Client.hs]: lol, this question doesn't make sense anymore

File: code/network/lib/Loot/Network/ZMQ/Client.hs (hunk 1 above)
Comment (volhovm) [reply code/network/lib/Loot/Network/ZMQ/Client.hs]: You're right. TBH I have no idea how "disconnecting peers" may be useful in DSCP scenario, so I never actually tested it.

File: code/network/lib/Loot/Network/ZMQ/Client.hs (hunk 3 above)
Comment (volhovm) [reply code/network/lib/Loot/Network/ZMQ/Client.hs]: I still think that reconnecting peers is something we should get rid of, implicitly offloading it to the ZMQ backend.

File: code/network/lib/Loot/Network/ZMQ/Client.hs (hunk 6 above)
Comment (volhovm) [reply code/network/lib/Loot/Network/ZMQ/Client.hs]: This is `DEALER`, so yes, the first frame must be `""`, read the guide.

File: code/network/lib/Loot/Network/ZMQ/Client.hs (hunk 8 above)
Comment (volhovm) [reply code/network/lib/Loot/Network/ZMQ/Client.hs]: Actually `lisetenersToWorker` is a copy-paste of `clientsToBroker`. Hmmm, it's hard to decide. I think generalising these two would only complicate the understanding. Workers are essentially three lines, but generalising them would require passing many parameters to the function. I guess I'd like to avoid that.

File: code/network/lib/Loot/Network/ZMQ/Client.hs (hunk 9 above)
Comment (volhovm) [reply code/network/lib/Loot/Network/ZMQ/Client.hs]: That's definitely not a healthy urge.

File: code/network/lib/Loot/Network/ZMQ/Client.hs (hunk 10 above)
Comment (pva701) [reply code/network/lib/Loot/Network/ZMQ/Client.hs]: lool, speed, it's because you join all client workers to one thread, soo speedy

Comment (pva701) [reply code/network/lib/Loot/Network/ZMQ/Client.hs]: sure there is place where you can compute length one time and reuse it, before `action` funtion

File: code/network/lib/Loot/Network/ZMQ/Client.hs (hunk 9 above)
Comment (pva701) [reply code/network/lib/Loot/Network/ZMQ/Client.hs]: lool, look at this code: some insane indexes which only you understand and only now, it's easier to find a bug in one function then in 4 compares among a lot of another code

File: code/network/lib/Loot/Network/ZMQ/Client.hs (hunk 10 above)
Comment (volhovm) [reply code/network/lib/Loot/Network/ZMQ/Client.hs]: >you join all client workers to one thread

Computing `cpdPolls` requires an STM action each poll, so yes, it's a noticeable overhead I believe.

>sure there is place where you can compute length one time and reuse it, before action funtion

Our peers list is not constant, it evolves over time.

Comment (pva701) [reply code/network/lib/Loot/Network/ZMQ/Client.hs]: I believe that one iteration of 
`forM_ (events `zip` [0..]) $ \(e,i) ->`
takes more time than performing `length cpdPolls` after reading it from `TVar`, however, I don't want to dispute about it. let's it be in this way

File: code/network/lib/Loot/Network/ZMQ/Client.hs (hunk 6 above)
Comment (pva701) [reply code/network/lib/Loot/Network/ZMQ/Client.hs]: >This is DEALER, so yes, the first frame must be "", read the guide.

ok, I will. but aside the guide: which motivation?

File: code/network/lib/Loot/Network/ZMQ/Client.hs (hunk 10 above)
Comment (pva701) [reply code/network/lib/Loot/Network/ZMQ/Client.hs]: >I believe that one iteration of
forM_ (events zip [0..]) $ \(e,i) ->
takes more time than performing length cpdPolls

do you realise it?

File: code/network/lib/Loot/Network/ZMQ/Client.hs (hunk 9 above)
Comment (pva701) [reply code/network/lib/Loot/Network/ZMQ/Client.hs]: ne ne, it's not resolved

Comment (pva701) [reply code/network/lib/Loot/Network/ZMQ/Client.hs]: I can do it myself in neighbour PR, but I insist it has to be done, because these checks
```
if | i < cpdN ->
| i < 2 * cpdN ->
| i == 2 * cpdN ->
i == 2 * cpdN + 1
```
really hard to follow, and I prefer to have it this way
```
case classifySockIndex i cpdN where
    BackSock peerId ->
    SubSock peerId   ->
    CliSock                 ->
    ReqSock               ->
```

Comment (pva701) [reply code/network/lib/Loot/Network/ZMQ/Client.hs]: so if you are interested how I recognise what each brunch does: "taaks ebana, chto tut u nas v tele, aga `receiveSub`, pohodu eto sub socket", so until I look at body I can't recognise the case.
